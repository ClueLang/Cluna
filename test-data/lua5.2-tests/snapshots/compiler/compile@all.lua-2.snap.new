---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
