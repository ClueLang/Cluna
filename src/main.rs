#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
