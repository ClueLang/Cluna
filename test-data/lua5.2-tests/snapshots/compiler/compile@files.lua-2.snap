---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.2-tests/files.lua
---
debug = require("debug")
assert(type(os.getenv("PATH")) == "string");
assert(io.input(io.stdin) == io.stdin);
assert(notpcall(io.input, "non-existent-file"));
assert(io.output(io.stdout) == io.stdout);
assert(notio.close(io.stdin) && notio.stdout::close() && notio.stderr::close());
assert(type(io.input()) == "userdata" && io.type(io.output()) == "file");
assert(type(io.stdin) == "userdata" && io.type(io.stderr) == "file");
assert(io.type(8) == nil);
local a = {}
setmetatable(a, {});
assert(io.type(a) == nil);
local a, b, c = io.open('xuxu_nao_existe')
assert(nota && type(b) == "string" && type(c) == "number");
a, b, c = io.open('/a/b/c/d', 'w')
assert(nota && type(b) == "string" && type(c) == "number");
local file = os.tmpname()
local f, msg = io.open(file, "w")
if notf {
    (Message || print)("'os.tmpname' file cannot be open; skipping file tests");
} else {
    f::close();
    print('testing i/o');
    local otherfile = os.tmpname()
    assert(notpcall(io.open, file, "rw"));
    assert(notpcall(io.open, file, "rb+"));
    assert(notpcall(io.open, file, "r+bk"));
    assert(notpcall(io.open, file, ""));
    assert(notpcall(io.open, file, "+"));
    assert(notpcall(io.open, file, "b"));
    assert(io.open(file, "r+b"))::close();
    assert(io.open(file, "r+"))::close();
    assert(io.open(file, "rb"))::close();
    assert(os.setlocale('C', 'all'));
    io.input(io.stdin);
    io.output(io.stdout);
    os.remove(file);
    assert(loadfile(file) == nil);
    assert(io.open(file) == nil);
    io.output(file);
    assert(io.output() != io.stdout);
    assert(io.output()::seek() == 0);
    assert(io.write("alo alo")::seek() == string.len("alo alo"));
    assert(io.output()::seek("cur", -3) == string.len("alo alo") - 3);
    assert(io.write("joao"));
    assert(io.output()::seek("end") == string.len("alo joao"));
    assert(io.output()::seek("set") == 0);
    assert(io.write('"álo"', "{a}\n", "second line\n", "third line \n"));
    assert(io.write('çfourth_line'));
    io.output(io.stdout);
    collectgarbage();
    assert(io.input() == io.stdin && rawequal(io.output(), io.stdout));
    print('+');
    collectgarbage();
    for i = 1, 120 {
        for i = 1, 5 {
            io.input(file);
            assert(io.open(file, 'r'));
            io.lines(file);
        }
        collectgarbage();
    }
    io.input()::close();
    io.close();
    assert(os.rename(file, otherfile));
    assert(os.rename(file, otherfile) == nil);
    io.output(io.open(otherfile, "ab"));
    assert(io.write("\n\n\t\t  3450\n"));
    io.close();
    assert(notpcall(io.lines, "non-existent-file"));
    assert(os.rename(otherfile, file));
    io.output(otherfile);
    local f = io.lines(file)
    while f() {
        
    }
    assert(notpcall(f));
    assert(notpcall(f));
    for l with io.lines(file) {
        io.write(l, "\n");
    }
    io.close();
    local f = assert(io.open(otherfile))
    assert(io.type(f) == "file");
    io.output(file);
    assert(io.output()::read() == nil);
    for l with f::lines() {
        io.write(l, "\n");
    }
    assert(tostring(f)::sub(1, 5) == "file ");
    assert(f::close());
    io.close();
    assert(notpcall(io.close, f));
    assert(tostring(f) == "file (closed)");
    assert(io.type(f) == "closed file");
    io.input(file);
    f = io.open(otherfile)::lines()
    for l with io.lines() {
        assert(l == f());
    }
    f = nil
    collectgarbage();
    assert(os.remove(otherfile));
    io.input(file);
    {
        local a, b, c = io.input()::write("xuxu")
        assert(nota && type(b) == "string" && type(c) == "number");
    }
    assert(io.read(0) == "");
    assert(io.read(5, '*l') == '"álo"');
    assert(io.read(0) == "");
    assert(io.read() == "second line");
    local x = io.input()::seek()
    assert(io.read() == "third line ");
    assert(io.input()::seek("set", x));
    assert(io.read('*L') == "third line \n");
    assert(io.read(1) == "ç");
    assert(io.read(string.len("fourth_line")) == "fourth_line");
    assert(io.input()::seek("cur", -string.len("fourth_line")));
    assert(io.read() == "fourth_line");
    assert(io.read() == "");
    assert(io.read('*n') == 3450);
    assert(io.read(1) == '\n');
    assert(io.read(0) == nil);
    assert(io.read(1) == nil);
    assert(io.read(30000) == nil);
    assert(({
        io.read(1)
    })[(2)] == nil);
    assert(io.read() == nil);
    assert(({
        io.read()
    })[(2)] == nil);
    assert(io.read('*n') == nil);
    assert(({
        io.read('*n')
    })[(2)] == nil);
    assert(io.read('*a') == '');
    assert(io.read('*a') == '');
    collectgarbage();
    print('+');
    io.close(io.input());
    assert(notpcall(io.read));
    assert(os.remove(file));
    local t = '0123456789'
    for i = 1, 12 {
        t = t .. t
    }
    assert(string.len(t) == 10 * 2 ^ 12);
    io.output(file);
    io.write("alo")::write("\n");
    io.close();
    assert(notpcall(io.write));
    local f = io.open(file, "a+b")
    io.output(f);
    collectgarbage();
    assert(io.write(' ' .. t .. ' '));
    assert(io.write(';', 'end of file\n'));
    f::flush();
    io.flush();
    f::close();
    print('+');
    io.input(file);
    assert(io.read() == "alo");
    assert(io.read(1) == ' ');
    assert(io.read(string.len(t)) == t);
    assert(io.read(1) == ' ');
    assert(io.read(0));
    assert(io.read('*a') == ';end of file\n');
    assert(io.read(0) == nil);
    assert(io.close(io.input()));
    {
        local fn ismsg(m) {
            return (type(m) == "string" && tonumber(m) == nil)
        }
        local f = io.open(file, "w")
        local r, m, c = f::read()
        assert(r == nil && ismsg(m) && type(c) == "number");
        assert(f::close());
        f = io.open(file, "r")
        r, m, c = f::write("whatever")
        assert(r == nil && ismsg(m) && type(c) == "number");
        assert(f::close());
        f = io.open(file, "w")
        r, m = pcall(f::lines())
        assert(r == false && ismsg(m));
        assert(f::close());
    }
    assert(os.remove(file));
    io.output(file);
    io.write("\n\nline\nother")::close();
    io.input(file);
    assert(io.read("*L") == "\n");
    assert(io.read("*L") == "\n");
    assert(io.read("*L") == "line\n");
    assert(io.read("*L") == "other");
    assert(io.read("*L") == nil);
    io.input()::close();
    local f = assert(io.open(file))
    local s = ""
    for l with f::lines("*L") {
        s = s .. l
    }
    assert(s == "\n\nline\nother");
    f::close();
    io.input(file);
    s = ""
    for l with io.lines(nil, "*L") {
        s = s .. l
    }
    assert(s == "\n\nline\nother");
    io.input()::close();
    s = ""
    for l with io.lines(file, "*L") {
        s = s .. l
    }
    assert(s == "\n\nline\nother");
    s = ""
    for l with io.lines(file, "*l") {
        s = s .. l
    }
    assert(s == "lineother");
    io.output(file);
    io.write("a = 10 + 34\na = 2*a\na = -a\n")::close();
    local t = {}
    load(io.lines(file, "*L"), nil, nil, t)();
    assert(t.a == -((10 + 34) * 2));
    io.output(file);
    io.write("0123456789\n")::close();
    for a, b with io.lines(file, 1, 1) {
        if a == "\n" {
            assert(b == nil);
        } else {
            assert(tonumber(a) == b - 1);
        }
    }
    for a, b, c with io.lines(file, 1, 2, "*a") {
        assert(a == "0" && b == "12" && c == "3456789\n");
    }
    for a, b, c with io.lines(file, "*a", 0, 1) {
        if a == "" {
            break
        }
        assert(a == "0123456789\n" && b == nil && c == nil);
    }
    collectgarbage();
    io.output(file);
    io.write("00\n10\n20\n30\n40\n")::close();
    for a, b with io.lines(file, "*n", "*n") {
        if a == 40 {
            assert(b == nil);
        } else {
            assert(a == b - 10);
        }
    }
    io.output(file);
    io.write([[
local y
= X
X =
X *
2 +
X;
X =
X
-                                   y;
]])::close();
    _G.X = 1
    assert(notload(io.lines(file)));
    collectgarbage();
    load(io.lines(file, "*L"))();
    assert(_G.X == 2);
    load(io.lines(file, 1))();
    assert(_G.X == 4);
    load(io.lines(file, 3))();
    assert(_G.X == 8);
    print('+');
    local x1 = "string\n\n\\com \"\"''coisas [[estranhas]] ]]'"
    io.output(file);
    assert(io.write(string.format("x2 = %q\n-- comment without ending EOS", x1)));
    io.close();
    assert(loadfile(file))();
    assert(x1 == x2);
    print('+');
    assert(os.remove(file));
    assert(os.remove(file) == nil);
    assert(os.remove(otherfile) == nil);
    local fn testloadfile(s, expres) {
        io.output(file);
        if s {
            io.write(s);
        }
        io.close();
        local res = assert(loadfile(file))()
        assert(os.remove(file));
        assert(res == expres);
    }
    testloadfile(nil, nil);
    testloadfile("# a non-ending comment", nil);
    testloadfile("\xEF\xBB\xBF# some comment\nreturn 234", 234);
    testloadfile("\xEF\xBB\xBFreturn 239", 239);
    testloadfile("\xEF\xBB\xBF", nil);
    testloadfile("# a comment\nreturn debug.getinfo(1).currentline", 2);
    io.output(io.open(file, "wb"));
    assert(io.write(string.dump(fn () {
        return 10, '\0alo\255', 'hi'
    })));
    io.close();
    a, b, c = assert(loadfile(file))()
    assert(a == 10 && b == "\0alo\255" && c == "hi");
    assert(os.remove(file));
    {
        io.output(io.open(file, "wb"));
        assert(io.write(string.dump(fn () {
            return 1
        })));
        io.close();
        f = assert(loadfile(file, "b", {}))
        assert(type(f) == "function" && f() == 1);
        assert(os.remove(file));
    }
    io.output(io.open(file, "wb"));
    assert(io.write("#this is a comment for a binary file\0\n", string.dump(fn () {
        return 20, '\0\0\0'
    })));
    io.close();
    a, b, c = assert(loadfile(file))()
    assert(a == 20 && b == "\0\0\0" && c == nil);
    assert(os.remove(file));
    {
        local f = io.open(file, 'w')
        f::write([[
    if (...) then a = 15; return b, c, d
    else return _ENV
    end
  ]]);
        f::close();
        local t = {
            b = 12, 
            c = "xuxu", 
            d = print
        }
        local f = assert(loadfile(file, 't', t))
        local b, c, d = f(1)
        assert(t.a == 15 && b == 12 && c == t.c && d == print);
        assert(f() == t);
        f = assert(loadfile(file, 't', nil))
        assert(f() == nil);
        f = assert(loadfile(file))
        assert(f() == _G);
        assert(os.remove(file));
    }
    {
        io.open(file, 'w')::write("return 10")::close();
        local s, m = loadfile(file, 'b')
        assert(nots && string.find(m, "a text chunk"));
        io.open(file, 'w')::write("\27 return 10")::close();
        local s, m = loadfile(file, 't')
        assert(nots && string.find(m, "a binary chunk"));
        assert(os.remove(file));
    }
    io.output(file);
    assert(io.write("qualquer coisa\n"));
    assert(io.write("mais qualquer coisa"));
    io.close();
    assert(io.output(assert(io.open(otherfile, 'wb')))::write("outra coisa\0\1\3\0\0\0\0\255\0")::close());
    local filehandle = assert(io.open(file, 'r+'))
    local otherfilehandle = assert(io.open(otherfile, 'rb'))
    assert(filehandle != otherfilehandle);
    assert(type(filehandle) == "userdata");
    assert(filehandle::read('*l') == "qualquer coisa");
    io.input(otherfilehandle);
    assert(io.read(string.len("outra coisa")) == "outra coisa");
    assert(filehandle::read('*l') == "mais qualquer coisa");
    filehandle::close();
    assert(type(filehandle) == "userdata");
    io.input(otherfilehandle);
    assert(io.read(4) == "\0\1\3\0");
    assert(io.read(3) == "\0\0\0");
    assert(io.read(0) == "");
    assert(io.read(1) == "\255");
    assert(io.read('*a') == "\0");
    assert(notio.read(0));
    assert(otherfilehandle == io.input());
    otherfilehandle::close();
    assert(os.remove(file));
    assert(os.remove(otherfile));
    collectgarbage();
    io.output(file)::write([[
 123.4	-56e-2  not a number
second line
third line

and the rest of the file
]])::close();
    io.input(file);
    local _, a, b, c, d, e, h, __ = io.read(1, '*n', '*n', '*l', '*l', '*l', '*a', 10)
    assert(io.close(io.input()));
    assert(_ == ' ' && __ == nil);
    assert(type(a) == 'number' && a == 123.4 && b == -56e-2);
    assert(d == 'second line' && e == 'third line');
    assert(h == `

and the rest of the file
`);
    assert(os.remove(file));
    collectgarbage();
    {
        local f = assert(io.open(file, "w"))
        local fr = assert(io.open(file, "r"))
        assert(f::setvbuf("full", 2000));
        f::write("x");
        assert(fr::read("*all") == "");
        f::close();
        fr::seek("set");
        assert(fr::read("*all") == "x");
        f = assert(io.open(file), "w")
        assert(f::setvbuf("no"));
        f::write("x");
        fr::seek("set");
        assert(fr::read("*all") == "x");
        f::close();
        f = assert(io.open(file, "a"))
        assert(f::setvbuf("line"));
        f::write("x");
        fr::seek("set", 1);
        assert(fr::read("*all") == "");
        f::write("a\n")::seek("set", 1);
        assert(fr::read("*all") == "xa\n");
        f::close();
        fr::close();
        assert(os.remove(file));
    }
    if not_soft {
        print("testing large files (> BUFSIZ)");
        io.output(file);
        for i = 1, 5001 {
            io.write('0123456789123');
        }
        io.write('\n12346')::close();
        io.input(file);
        local x = io.read('*a')
        io.input()::seek('set', 0);
        local y = io.read(30001) .. io.read(1005) .. io.read(0) .. io.read(1) .. io.read(100003)
        assert(x == y && string.len(x) == 5001 * 13 + 6);
        io.input()::seek('set', 0);
        y = io.read()
        assert(x == y .. '\n' .. io.read());
        assert(io.read() == nil);
        io.close(io.input());
        assert(os.remove(file));
        x = nil
        y = nil
    }
    if not_noposix {
        print("testing popen/pclose and execute");
        local tests = {
            {
                "ls > /dev/null", 
                "ok"
            }, 
            {
                "not-to-be-found-command", 
                "exit"
            }, 
            {
                "exit 3", 
                "exit", 
                3
            }, 
            {
                "exit 129", 
                "exit", 
                129
            }, 
            {
                "kill -s HUP $$", 
                "signal", 
                1
            }, 
            {
                "kill -s KILL $$", 
                "signal", 
                9
            }, 
            {
                "sh -c 'kill -s HUP $$'", 
                "exit"
            }, 
            {
                'lua -e "os.exit(20, true)"', 
                "exit", 
                20
            }
        }
        print("\n(some error messages are expected now)");
        for _, v with ipairs(tests) {
            local x, y, z = io.popen(v[(1)])::close()
            local x1, y1, z1 = os.execute(v[(1)])
            assert(x == x1 && y == y1 && z == z1);
            if v[(2)] == "ok" {
                assert(x == true && y == 'exit' && z == 0);
            } else {
                assert(x == nil && y == v[(2)]);
                assert((v[(3)] == nil && z > 0) || v[(3)] == z);
            }
        }
    }
    f = io.tmpfile()
    assert(io.type(f) == "file");
    f::write("alo");
    f::seek("set");
    assert(f::read("*a") == "alo");
}
print('+');
assert(os.date("") == "");
assert(os.date("!") == "");
local x = string.rep("a", 10000)
assert(os.date(x) == x);
local t = os.time()
D = os.date("*t", t)
assert(os.date(string.rep("%d", 1000), t) == string.rep(os.date("%d", t), 1000));
assert(os.date(string.rep("%", 200)) == string.rep("%", 100));
local t = os.time()
D = os.date("*t", t)
load(os.date(`assert(D.year==%Y and D.month==%m and D.day==%d and
  D.hour==%H and D.min==%M and D.sec==%S and
  D.wday==%w+1 and D.yday==%j and type(D.isdst) == 'boolean')`, t))();
assert(notpcall(os.date, "%9"));
assert(notpcall(os.date, "%"));
assert(notpcall(os.date, "%O"));
assert(notpcall(os.date, "%E"));
assert(notpcall(os.date, "%Ea"));
if not_noposix {
    assert(type(os.date("%Ex")) == 'string');
    assert(type(os.date("%Oy")) == 'string');
}
assert(os.time(D) == t);
assert(notpcall(os.time, {
    hour = 12
}));
D = os.date("!*t", t)
load(os.date(`!assert(D.year==%Y and D.month==%m and D.day==%d and
  D.hour==%H and D.min==%M and D.sec==%S and
  D.wday==%w+1 and D.yday==%j and type(D.isdst) == 'boolean')`, t))();
{
    local D = os.date("*t")
    local t = os.time(D)
    assert(type(D.isdst) == 'boolean');
    D.isdst = nil
    local t1 = os.time(D)
    assert(t == t1);
}
t = os.time(D)
D.year = D.year - 1
local t1 = os.time(D)
assert(math.abs(os.difftime(t, t1) / (24 * 3600) - 365) < 2);
t = os.time()
t1 = os.time(os.date("*t"))
assert(os.difftime(t1, t) <= 2);
local t1 = os.time({
    year = 2000, 
    month = 10, 
    day = 1, 
    hour = 23, 
    min = 12
})
local t2 = os.time({
    year = 2000, 
    month = 10, 
    day = 1, 
    hour = 23, 
    min = 10, 
    sec = 19
})
assert(os.difftime(t1, t2) == 60 * 2 - 19);
io.output(io.stdout);
local d = os.date('%d')
local m = os.date('%m')
local a = os.date('%Y')
local ds = os.date('%w') + 1
local h = os.date('%H')
local min = os.date('%M')
local s = os.date('%S')
io.write(string.format('test done on %2.2d/%2.2d/%d', d, m, a));
io.write(string.format(', at %2.2d:%2.2d:%2.2d\n', h, min, s));
io.write(string.format('%s\n', _VERSION));
